Improve implementation of quest items

pedro-w opened the pull request:
The Quest UI (accessible from the 'RETURN' menu)  now shows a selectable list of item names on the left and corresponding text on the right. You can press ALT to switch to the right hand pane and scroll up/down if there is more text than will fit in the box.

However I removed all the item definitions from the scripts (they were just placeholder text) so currently it is not possible to see this view - hopefully I can start to add some items in before 1.0 is ready

Fixes #125

pedro-w committed:
Improve UI for Quest items

At the moment there are no items, so selecting that option
just gives an 'ow' and does nothing.

pedro-w committed:
Allow scrolling of long text items

pedro-w committed:
Rename for consistency

pedro-w commented:
![Screenshot 2022-08-11 at 13 32 48](https://user-images.githubusercontent.com/22172519/184134169-a7076fea-ccf1-4bfe-96a5-bdda61f6918e.png)
FYI this is what it looks like

z9484 commented on scripts/global.lua:
```diff
@@ -542,13 +542,11 @@ end
-   add_quest_item(_"Test1",    _"Some test info")
-   add_quest_item(_"Sensar",   _"He rages!")
+   -- Add global quest items here:
+   -- add_quest_item("key", "text")
```
Did the format change? I uncommented this out and it gives a seg fault when you go to the quest menu. It seems to crash on the layout menu.cpp:395

pedro-w replied:
I couldn't reproduce this. Did you just uncomment line 549, ie.
``` lua
function get_quest_info()
   if LOC_get_quest_info then
      LOC_get_quest_info()
   end
   -- Add global quest items here:
   add_quest_item("key", "text")
end
```
Can you give more info on the seg fault?

pedro-w replied:
Fixed now, MSVC does extra checking on iterators and that was causing an abort

OnlineCop commented:
After pulling this, I played through till I got to town2; I kept checking Quest but it never populated. I also searched for the text "Look for Derig in the grotto" but didn't find it.

Do I have to cheat those quests in, or do something specific to trigger the quests into my Quest list?

pedro-w commented:
> Do I have to cheat those quests in, or do something specific to trigger the quests into my Quest list?

I should have been clearer - in the top comment I said "However I removed all the item definitions from the scripts (they were just placeholder text) so currently it is not possible to see this view" - that's why I put the screenshot in just as an example of what it *would* look like.
If you want to see it in action you could add some calls to `add_quest_item` in global.lua,  function `get_quest_info` (line 550-ish)

pedro-w commented:
I'm going to improve this, to give some feedback if there's a script error rather than just silently not working which is what happens now. So it's not ready for merge yet.
Still haven't been able to reproduce your segfault, Z, maybe it's a windows thing?

pedro-w committed:
Use protected call for get_quest_info

Also put the call logic into a separate function and re-use it
for autoexec, postexec, zone_handler and entity_handler

z9484 commented:
Yeah maybe its windows only. Heres where it crashes.
![Screenshot 2022-08-15 quest](https://user-images.githubusercontent.com/87368/184774135-cf5d509f-004a-4d07-8061-7f1f3d836c5f.png)

pedro-w commented:
it may be I made an incorrect assumption that just happened to work on Mac - I'll start up my old W10 laptop

pedro-w commented:
Yow.
![yow](https://user-images.githubusercontent.com/22172519/184954932-87739523-a94c-46de-badb-4674382256f9.png)

pedro-w committed:
Fix crash when using MSVC's C++ Standard Library

pedro-w commented:
OK I think I have fixed it now - on Mac and Linux it's OK to form an out-of-range string iterator as long as you don't ever dereference it, but MSVC complains (at least in debug mode)

OnlineCop commented:
This is what I see when running from Visual Studio 2022:

![image](https://user-images.githubusercontent.com/1168697/184969905-98b08b71-d637-4b79-9426-b59d7c0c81b3.png)

Would it look good if the circled chevrons were closer to the right edge of the text window?

This chevron is aligned with the right edge of its menu when there are enough quests to scroll to another page:

![image](https://user-images.githubusercontent.com/1168697/184970382-1dfa7f24-a614-4184-9f94-ffffb2990eab.png)

And this is entirely up to you: are there any other colors that we could use with the chevrons to make them stand out a little easier against the white edge of the menus?

OnlineCop commented:
I haven't looked over the word-wrap code very carefully, but I threw together a bunch of edge cases to consider (punctuation, super long text, etc.

Super impressed.

![image](https://user-images.githubusercontent.com/1168697/184981946-52b5e124-14c1-4699-ad35-03425961c655.png)

Everything I worried would cause visual anomalies seem to be handled just fine.

Remember to clang-format the code before you merge 😉.

OnlineCop commented on src/menu.cpp:
```diff
@@ -272,11 +273,66 @@ void KMenu::menu()
+{
+    return it == end || isspace(*it);
+}
+static std::string::const_iterator cut(std::string::const_iterator from, std::string::const_iterator end)
```
Having comments for the parameters would be nice throughout, if possible.

pedro-w replied:
done

pedro-w commented:
> Would it look good if the circled chevrons were closer to the right edge of the text window?

![Screenshot 2022-08-16 at 21 51 00](https://user-images.githubusercontent.com/22172519/184983383-28c113a3-e062-405e-a98b-411cca88558a.png)

> And this is entirely up to you: are there any other colors that we could use with the chevrons to make them stand out a little easier against the white edge of the menus?

They're graphics from misc.png, there is a 'gold' right chevron but not up & down. I suppose they could be added. At the moment it's using the graphics `upptr` and `dnptr` which are defined elsewhere. Definitely do-able.

OnlineCop reviewed (approved):
Word wrap in this seems to work quite well. I like that you considered both the entries, and the descriptive text for them, to spill over the visible area and need the chevrons to indicate more text above/below.

pedro-w commented:
> Everything I worried would cause visual anomalies seem to be handled just fine.

I think it's ok, sometimes it looks like it could fit an extra word onto the line but if you count, it couldn't. The RHS looks raggedy because there's only 20 big characters per line. I guess it would be possible to stretch the spaces but that seemed too hard!
 
> Remember to clang-format the code before you merge 😉.

Sure

OnlineCop commented:
Is there a limit on the length of the quest keys ("New Quest 11" was a little too wide to fit in the left menu) that we need to keep in mind? We can avoid extra logic if you just throw a comment into global.lua warning the user to keep it under 11 characters.

OnlineCop commented:
Do you think that this word wrap could also be incorporated into (and/or replace) `KDraw::relay()`, since that too seems to be doing its own word wrapping?

pedro-w commented:
I wrote it with the idea that it could one day replace relay() but I didn't
want to mess about with code that was working already. Note that relay()
allows manual line breaks with the \n character but this new code doesn't
(yet, it would be easy enough to add)


On Tue, 16 Aug 2022 at 23:22, OnlineCop ***@***.***> wrote:

> Do you think that this word wrap could also be incorporated into (and/or
> replace) KDraw::relay(), since that too seems to be doing its own word
> wrapping?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/OnlineCop/kq-fork/pull/137#issuecomment-1217229037>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFJFGZ4NO2BTEXXKTZ77YBLVZQIA7ANCNFSM56HAYERQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

pedro-w committed:
Move text layout to KDraw

pedro-w committed:
Add comments and format source

pedro-w committed:
Move position of up/down arrows

pedro-w commented:
> Do you think that this word wrap could also be incorporated into (and/or replace) `KDraw::relay()`, since that too seems to be doing its own word wrapping?

I moved the layout code into KDraw, it made more sense there. I have prepared some code to replace `KDraw::relay` and it works OK but maybe will defer it until after 'PullZilla' (#142 and friends) :wink:
